### PR TITLE
Move unique functionality into getGroups to reduce calls to google

### PR DESCRIPTION
**Pull request by snuggie12**
#### Overview

Removed the unique function and integrated into `getGroups` to reduce the number of calls to google.

#### What this PR does / why we need it

Depending on the groups a user belongs to the current code can make several extraneous calls to google. This can occur either when a user has several indirect memberships to a group or when the user has both an indirect and direct membership to a group:

```
# two extraneous calls to numbers
# email of user/group -> [groups email is a member of]
user -> [one, two, three]
one -> [numbers]
two -> [numbers]
three -> [numbers]
numbers -> []

# extraneous call to bar due to both direct and indirect membership
user -> [foo, bar]
foo -> [bar]
bar -> []
```

Implementing a standard map setup reduces calls to google down to `number of groups + original user`.

I'm not sure if google allows it, but it also eliminates a stack overflow if there was a circular membership:
```
user -> [one]
one -> [two]
two -> [three]
three -> [one]
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by snuggie12** [comment]
@nabokihms do I add a label or do you all? It wasn't mentioned in the PR template, but wasn't sure if you would get notified of the PR until it's there

File: connector/google/google.go [hunk 1]
```diff
@@ -257,12 +258,15 @@ func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership
...
+			if _, exists := (*checkedGroups)[group.Email]; !exists {
+				(*checkedGroups)[group.Email] = struct{}{}
+				// TODO (joelspeed): Make desired group key configurable
+				userGroups = append(userGroups, group.Email)
+			}
 
 			// getGroups takes a user's email/alias as well as a group's email/alias
 			if fetchTransitiveGroupMembership {
-				transitiveGroups, err := c.getGroups(group.Email, fetchTransitiveGroupMembership)
+				transitiveGroups, err := c.getGroups(group.Email, fetchTransitiveGroupMembership, checkedGroups)
```
Comment (nabokihms) [inline connector/google/google.go:269]: Do we need to check transitive group membership if a groups is already in checked groups?

**Review by nabokihms — CHANGES_REQUESTED** [review]
Thanks for opening this PR. I left some questions. Could you please have a look?

File: connector/google/google.go (hunk 1 above)
Comment (snuggie12) [reply connector/google/google.go:269]: I'm unsure if I understand the question, but if you are asking if `fetchTransitiveGroupMembership` should always be `false` here then I think the answer is no due to multiple levels of nesting.

File: connector/google/google.go [hunk 2]
```diff
@@ -257,12 +258,15 @@ func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership
 		}
 
 		for _, group := range groupsList.Groups {
-			// TODO (joelspeed): Make desired group key configurable
-			userGroups = append(userGroups, group.Email)
+			if _, exists := (*checkedGroups)[group.Email]; !exists {
```
[the remark this answers is not in the document]
Comment (snuggie12) [reply connector/google/google.go:261]: Yeah, I would like to add a test. Would I need to setup a mock for that? Is there an example?

Locally I have logs which demonstrate this. I could add that to the code instead.

File: connector/google/google.go [hunk 3]
```diff
@@ -245,7 +246,7 @@ func (c *googleConnector) createIdentity(ctx context.Context, identity connector
 
 // getGroups creates a connection to the admin directory service and lists
 // all groups the user is a member of
-func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership bool) ([]string, error) {
+func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership bool, checkedGroups *map[string]struct{}) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (snuggie12) [reply connector/google/google.go:249]: Will do

File: connector/google/google.go (hunk 1 above)
Comment (snuggie12) [reply connector/google/google.go:269]: I see what you were saying now. I've moved this block inside of the if condition as well.

File: connector/google/google.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (snuggie12) [reply connector/google/google.go]: I believe this is done. Do you resolve or do I?

File: connector/google/google.go (hunk 1 above)
Comment (snuggie12) [reply connector/google/google.go]: Done. Same question as above on who resolves

File: connector/google/google.go [hunk 4]
```diff
@@ -257,17 +258,20 @@ func (c *googleConnector) getGroups(email string, fetchTransitiveGroupMembership
...
 		for _, group := range groupsList.Groups {
-			// TODO (joelspeed): Make desired group key configurable
-			userGroups = append(userGroups, group.Email)
-
-			// getGroups takes a user's email/alias as well as a group's email/alias
-			if fetchTransitiveGroupMembership {
-				transitiveGroups, err := c.getGroups(group.Email, fetchTransitiveGroupMembership)
-				if err != nil {
-					return nil, fmt.Errorf("could not list transitive groups: %v", err)
+			if _, exists := checkedGroups[group.Email]; !exists {
```
Comment (nabokihms) [inline connector/google/google.go:261]: NIT: I'd like to reduce the indentation level here, e.g., 

```go
if _, exists := checkedGroups[group.Email]; exists {
    continue
}
...
if !fetchTransitiveGroupMembership {
    continue
}
...    
```

Nested conditions wrapped in two for loops are a little bit harder to read.

File: connector/google/google.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply connector/google/google.go]: Oh, there are no tests for Google connector at all...

You can use [this code](https://github.com/dexidp/dex/blob/master/connector/bitbucketcloud/bitbucketcloud_test.go) as an example to write a simple test.

**Review by nabokihms — CHANGES_REQUESTED** [review]
Looks nice! Things that stop us from merging this PR:
1. A test.
2. [DCO](https://github.com/apps/dco).

**Comment by snuggie12** [comment]
> Looks nice! Things that stop us from merging this PR:
> 
>     1. A test.
> 
>     2. [DCO](https://github.com/apps/dco).

I'll get on adding the test now. Had a production incident all day yesterday to handle but should be able to get to it the next couple of days.

For DCO, is that just you want the initial commit to have the `Signed-off-by` line like the 2nd commit currently has or is there something more to it that I missed from your link?

Do you mind if I add a few `debug` level log lines stating which email is being checked and later if the group exists? If yes, do you want me to use `else` or log and still use `continue`??

File: connector/google/google.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (snuggie12) [reply connector/google/google.go]: The example looked easy enough in that you could send to whatever URL you want but the google connector is using a library and so I couldn't figure out how to do it in a similar fashion.

I attempted both interfaces and embedded structs but couldn't get around the issue of `getGroups` always calling its version of `getGroupsList`. In the end I settled on `getGroups` taking a function as a parameter and that worked.

Hopefully that is okay, because otherwise I fear I might need a lot of hand holding to get something working (not a developer by trade.)

File: connector/google/google.go (hunk 4 above)
Comment (unaryops) [reply connector/google/google.go]: Perhaps use some tools in the build pipeline that do static analysis of cognitive and cyclomatic complexity.

File: connector/google/google.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ichbinfrog) [reply connector/google/google.go]: @snuggie12, I've implemented some unit testing for your PR on my fork: https://github.com/ichbinfrog/dex/blob/reduce-admin-api-calls/connector/google/google_test.go. 

Feel free to pull / take inspiration from the test file and reuse it since it also adds the required instrumentation for the amount of apiCalls.

TLDR: the basic gist of mocking google APIs is to inject a few client options which will force it to point to a local http server whose behavior you can control:
```go
conn.adminSrv, err = admin.NewService(
   context.Background(), 
   option.WithoutAuthentication(), 
   option.WithEndpoint(ts.URL),
)
```

[the remark this answers is not in the document]
Comment (snuggie12) [reply connector/google/google.go]: > @snuggie12, I've implemented some unit testing for your PR on my fork: https://github.com/ichbinfrog/dex/blob/reduce-admin-api-calls/connector/google/google_test.go.
> 
> Feel free to pull / take inspiration from the test file and reuse it since it also adds the required instrumentation for the amount of apiCalls.
> 
> TLDR: the basic gist of mocking google APIs is to inject a few client options which will force it to point to a local http server whose behavior you can control:
> 
> ```go
> conn.adminSrv, err = admin.NewService(
>    context.Background(), 
>    option.WithoutAuthentication(), 
>    option.WithEndpoint(ts.URL),
> )
> ```

@ichbinfrog thanks for this. I've been busy moving where I live but I'll def take a look soon.

**Comment by snuggie12** [comment]
@ichbinfrog I reverted my `google.go` prior to me adding tests where I changed how some methods worked. I get an error when I run `go test` but decided to check it in assuming there's more to it than that for it to work properly.

If this doesn't work I might need your help debugging the tests.

**Comment by snuggie12** [comment]
@nabokihms I don't see a way to restart that one job and it appears to be unrelated to my change. Can you restart or merge anyways?

**Comment by snuggie12** [comment]
@nabokihms was there anything else you needed? I'd like to get this merged as soon as possible as I think I've seen some security releases and my company is currently running a custom build to fix this problem.

**Comment by snuggie12** [comment]
@sagikazarmark sorry if you're the wrong person. Saw you have lots of commits and modified my milestone. Who should I ask for review on this? Really need this feature and AFAIK all things are signed-off.

**Review by nabokihms — APPROVED** [review]

**Comment by nabokihms** [comment]
@snuggie12 sorry for the delay, and thank you for all your work and patience.
